ramp-engine/ramp_engine/aws/api.py

```diff
@@ -148,21 +152,43 @@ def launch_ec2_instances(config, nb=1):
     if use_spot_instance:
         logger.info('Attempting to use spot instance.')
         now = datetime.utcnow() + timedelta(seconds=3)
-        request_wait = timedelta(minutes=10)
-        response = client.request_spot_instances(
-            AvailabilityZoneGroup=config[REGION_NAME_FIELD],
-            InstanceCount=nb,
-            LaunchSpecification={
-                'SecurityGroups': [security_group],
-                'ImageId': ami_image_id,
-                'InstanceType': instance_type,
-                'KeyName': key_name,
-            },
-            Type='one-time',
-            ValidFrom=now,
-            ValidUntil=(now + request_wait),
-        )
-
+        wait_minutes = WAIT_MINUTES
+        max_tries_to_connect = MAX_TRIES_TO_CONNECT
+        request_wait = timedelta(minutes=wait_minutes)
+        n_try = 0
+        response = None
+        while not(response) and (n_try < max_tries_to_connect):
+            try:
+                response = client.request_spot_instances(
+                    AvailabilityZoneGroup=config[REGION_NAME_FIELD],
+                    InstanceCount=nb,
+                    LaunchSpecification={
+                        'SecurityGroups': [security_group],
+                        'ImageId': ami_image_id,
+                        'InstanceType': instance_type,
+                        'KeyName': key_name,
+                    },
+                    Type='one-time',
+                    ValidFrom=now,
+                    ValidUntil=(now + request_wait),
+                )
+            except botocore.exceptions.ClientError as e:
+                n_try += 1
+                if n_try < max_tries_to_connect:
+                    # wait before you try again
+                    logger.warning('Not enough instances available: I am going'
+                                   f' to wait for {wait_minutes} minutes'
+                                   ' before trying again (this was'
+                                   f' {n_try} out of {max_tries_to_connect}'
+                                   ' tries to connect)')
+                    time.sleep(wait_minutes*60)
+                else:
+                    logger.error(f'Not enough instances available: {e}')
+                    return None,
+            except Exception as e:
+                # unknown error
+                logger.error(f'AWS worker error: {e}')
+                return None,
```

Comment:
Not a big fan of `while` loop because I feel it is harder to read what is hapening.
I propose to use `for/break/else` structure to make if a bit clearer but this is mainly nitpicking so feel free to ignore the comment if it does not seem clearer to you:

```suggestion
        for n_try in range(max_tries_to_connect):
            try:
                response = client.request_spot_instances(
                    AvailabilityZoneGroup=config[REGION_NAME_FIELD],
                    InstanceCount=nb,
                    LaunchSpecification={
                        'SecurityGroups': [security_group],
                        'ImageId': ami_image_id,
                        'InstanceType': instance_type,
                        'KeyName': key_name,
                    },
                    Type='one-time',
                    ValidFrom=now,
                    ValidUntil=(now + request_wait),
                )
                break
            except botocore.exceptions.ClientError as e:
                # wait before you try again
                logger.warning('Not enough instances available: I am going'
                               f' to wait for {wait_minutes} minutes'
                               ' before trying again (this was'
                               f' {n_try} out of {max_tries_to_connect}'
                               ' tries to connect)')
                time.sleep(wait_minutes*60)
            except Exception as e:
                # unknown error
                logger.error(f'AWS worker error: {e}')
                return None,
        else:
            # reached max retries
            logger.error(f'Not enough instances available: {e}')
            return None,
```